### PR TITLE
Add `Functor` trait

### PR DESCRIPTION
**Pull request by janmasrovira**

### CODE REVIEW

File: Stdlib/Trait/Functor/Monomorphic.juvix [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+import Stdlib.Trait.Functor.Polymorphic as Poly;
+
+trait
+type Functor (container elem : Type) :=
+  mkFunctor {
+    syntax iterator map {init := 0; range := 1};
+    map : (elem -> elem) -> container -> container
+  };
+
+open Functor public;
```
Comment (paulcadman) [inline Stdlib/Trait/Functor/Monomorphic.juvix:14]: The duplicate `open` can be removed?

File: Stdlib/Trait/Functor/Polymorphic.juvix [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+import Stdlib.Data.Unit open;
+
+trait
+type Functor (f : Type -> Type) :=
+  mkFunctor {
+    syntax iterator map {init := 0; range := 1};
+    map : {A B : Type} -> (A -> B) -> f A -> f B
+  };
+
+open Functor public;
```
Comment (paulcadman) [inline Stdlib/Trait/Functor/Polymorphic.juvix:13]: The duplicate open Functor public can be removed?

**Review by paulcadman — COMMENTED** [review]
We could also add a Functor instance to the `Result` type.

File: Stdlib/Trait/Functor/Polymorphic.juvix (hunk 2 above)
Comment (janmasrovira) [reply Stdlib/Trait/Functor/Polymorphic.juvix:13]: fixed

File: Stdlib/Trait/Functor/Monomorphic.juvix (hunk 1 above)
Comment (janmasrovira) [reply Stdlib/Trait/Functor/Monomorphic.juvix:14]: fixed

**Comment by janmasrovira** [comment]
> We could also add a Functor instance to the `Result` type.

Done

**Review by paulcadman — APPROVED** [review]

**Pull request merged**